tools/codegen/dest/gen_external_aten_fallbacks.py

```diff
@@ -0,0 +1,305 @@
+    'acos',
+    'acosh',
+    'asin',
+    'asinh',
+    'atan',
+    'atan2',
+    'atanh',
+    'baddbmm',
+    'bernoulli',
+    'binary_cross_entropy',
+    'binary_cross_entropy_backward',
+    'clamp',
+    'div',
+    'gather',
+    'ger',
+    'hardsigmoid',
+    'kthvalue',
+    'index_select',
+    'inverse',
+    'log',
+    'masked_select',
+    'maximum',
+    'minimum',
+    'pow',
+    'prod',
+    'nonzero',
+    'round',
+    'normal',
+    'std',
+    'take',
+    'topk',
+    'var',
+]
+
+def requires_backend_wrapper(f: ExternalBackendFunction) -> bool:
+    requires_lowering = not any(is_generic_dispatch_key(k) for k in f.native_function.dispatch)
+    has_xla_lowering = f.metadata is not None
+    in_denylist = any([re.match(frx, str(f.native_function.func.name)) for frx in _FN_DENYLIST_REGEX])
+    return not in_denylist and (requires_lowering or has_xla_lowering)
+
+def xla_tensor_creation_api(ret_name: str, ret: Return, device_param_name: str, *, tuple_idx: Optional[int] = None) -> str:
+    if ret.type == BaseType(BaseTy.Tensor) and not ret.is_write:
+        # Only raw Tensor (non-reference) returns need to go through the XLA tensor creation API.
+        # Tensor references can be returned directly, since they've already been converted to XLA tensors.
+        # See Note [Tensor Copy Returns]
+        bridge_api = 'CreateXlaTensor'
+    elif isinstance(ret.type, ListType) and ret.type.elem == BaseType(BaseTy.Tensor):
+        bridge_api = 'CreateXlaTensors'
+    else:
+        # for non tensor-types, there's no need to wrap the output in an xla bridge api.
+        return ret_name
+
+    out_name = 'x_result'
+    if tuple_idx is not None:
+        out_name = f"std::get<{tuple_idx}>(x_result)"
+
+    return f"bridge::{bridge_api}({out_name}, bridge::GetXlaDevice({device_param_name}))"
+
+
+
+# Generates aten_xla_type_default.h and aten_xla_type_default.cpp.
+#
+#   - This function registers external backend kernels, and also generates fallbacks to CPU.
+#     This is useful because pretty much all external backends (e.g. XLA)
+#     do not have full aten coverage.
+#     For operators not implemented by the external backend, our codegen
+#     will register these fallbacks instead.
+#   - Why do we generate fallback for ALL aten ops, including ops that
+#     external backends have already implemented?
+#     Many external backend kernels only work with specific input shapes,
+#     and are written to call into a cpu fallback when given inputs
+#     that they cannot handle.
+@dataclass(frozen=True)
+class GenExternalAtenFallback:
+    target: Union[
+        Literal[Target.NAMESPACED_DEFINITION],
+        Literal[Target.NAMESPACED_DECLARATION],
+        Literal[Target.REGISTRATION],
+    ]
+
+    @method_with_native_function
+    def __call__(self, g: Union[ExternalBackendFunctionsGroup, ExternalBackendFunction]) -> List[str]:
+
+        def gen_out_wrapper(g: ExternalBackendFunctionsGroup) -> Optional[str]:
```

Comment:
Why not make these methods on GenExternalAtenFallback? Would save you a level of indentation, and brings this file more in line with the prevailing style of dest modules.

Reply:
One reservation that I had was for generating out wrappers: given an `ExternalBackendFunctionsGroup`, `gen_unstructured_external()` can't really operate on each `ExternalBackendFunction` independently: the codegen is supposed to generate an `out` wrapper if the `out` ExternalBackendFunction doesn't have ExternalMetadata, and that logic needs to know the name of the `functional` op (in order to call it). When the functions are nested as closures like they are now, `gen_unstructured_external()` has access to the higher-scope `ExternalBackendFunctionsGroup`.

Locally I ended up just doing what you said, but I had to update the signature to `gen_unstructured_external(f: ExternalBackendFunction, g: ExternalBackendFunctionGroup)`, so it has all of the information it needs. I think it's an overall win for readability, since that level of indentation made things pretty confusing.

Reply:
(I'm actually not gonna make this change for now, and just land the PR as is. lmk if you think like the change though)

Reply:
Well, what you said here doesn't seem to apply to `gen_out_wrapper`. And it is totally reasonable to pass it in as an extra argument, that's what we do in other places.

I'm not *too* allergic to nested closures, but I know @bhosmer has been on my case about them in the past, and explicitly lifting out the closures does make it easier to reason about what information is propagated and what is not.

Reply:
Removing the nested closures in this file will also be that much easier after I finish moving the `gen_out_wrapper` logic out of this file as part of https://github.com/pytorch/pytorch/pull/56835. I can do that in a follow-up.